File: Data Structures & Algorithms/decode-ways/submission-4.cpp
class Solution {
public:
    int numDecodings(string s) {
        int N = s.length();
        vector<int> dp(N+1, 0);
        int prev = s[N-1]-'0';
        if(prev != 0) dp[N-1] = 1;
        dp[N]=1;
        for(int i=N-2;i>=0;i--){
            int num = s[i]-'0';
            if(num == 0){
                prev = 0;
                continue;
            }
            if((num == 2 && prev < 7) || (num == 1)){
                dp[i] = dp[i+1] + dp[i+2];
            }else{
                dp[i] = dp[i+1];
            }
            prev=num;
        }
        return dp[0];
    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-4.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        s, f = head, head
        while f and s:
            if f.next:
                f = f.next.next
            else:
                break
            
            s = s.next

        prev = None
        curr = s.next
        s.next = None
        
        while curr:
            nxt = curr.next
            curr.next = prev
            prev = curr
            curr = nxt
        
        rev_head = prev


        curr = head
        rev_curr = rev_head
        
        while curr and rev_curr:
            nxt = curr.next
            r_nxt = rev_curr.next

            curr.next = rev_curr
            rev_curr.next = nxt

            curr = nxt
            rev_curr = r_nxt

        


